jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpField.java

```diff
@@ -65,7 +65,7 @@ public HttpField(HttpHeader header, String name, String value)
             _name = _header.asString();
         else
             _name = Objects.requireNonNull(name, "name");
-        _value = Objects.requireNonNull(value);
+        _value = value != null ? value : "";
```

Comment:
1) the javadoc states that `value` must be non-null
2) since you changed your mind, that deserves an explanation, and maybe a comment

Reply:
> This brings up the question: why make this constructor `public` at all? Just make it `private` and have the other constructors delegate to it.

This PR has been committed already, but I will try to make this constructor private in my related PRs.

Reply:
> it is used frequently to copy another fields name regardless of enum. We could do `HttpField(HttpField useNames, String value)` but that is ugly as well?

I would add a new method for this use-case: `withValue(newValue)` which would return a new instance with the same name but new value. If `value` is unchanged, you could return `this`.

Another approach would be ro modify `getName()` to return the field's name regardless of whether it comes from an `HttpHeader` or `_name`. Then to copy another field's name regardless of enum you would invoke `new HttpField(other.getName(), newValue)`.

I personally prefer the first option.

Reply:
I had another go at this and I think it is an ugliness that we have (and will live with) due to efficiency.

For example, in the various implementations of HTTP, we will often parse out a header name and then look up any associated `HttpHeader` for it to know if we need to act specially.   Now we have both the `HttpHeader` and name, so it would be silly to pass just one of them into the `HttpField` constructor so that it has to look up the other again.   I know it is a tiny cost, but it is in on the hot path and executed many times per request, so collectively it is run many many many times every second of every day.

Reply:
@cowwoc, the problem is how do you get the HttpField in the first place on which to call `withValue`.
It is possible to restructure our parsers to create new HttpField instances rather than call the cache themselves, but then we might be double handling the value.    
It is something that I will look at, but for now it is not the highest priority.

Reply:
> @cowwoc, the problem is how do you get the HttpField in the first place on which to call withValue.

How do you do it now?

Reply:
@cowwoc Here is an example from HTTP/2 header decoding where we already know the name and optional HttpHeader and use them to  make the right type of HttpField (including supporting optimised int and long values):
https://github.com/eclipse/jetty.project/blob/f4b97fbfb86c6f357adf7adbf95d3149a15e1e8c/jetty-core/jetty-http2/jetty-http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/HpackDecoder.java#L240-L272

Here is an similar example from HTTP/1 where we have again parsed the name and know the HttpHeader, but also need to consult a compliance mode to see if we can standardise the case insensitive name or if we have to preserve the case of the name as received:
https://github.com/eclipse/jetty.project/blob/f4b97fbfb86c6f357adf7adbf95d3149a15e1e8c/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java#L1104

It is never a simple as you'd think/like/hope

Reply:
Thank you, I understand what you mean now.

So, yes, the question is whether we can delegate the cache lookup to `HttpField` instead of looking up the values externally and passing them in.

At first glance, this seems to be possible. Couldn't you remove the external name and header fields and add an `HttpField` in their place?

As far as I can tell, every case can be replaced by a call to the `HttpField(HttpHeader, String)` or `HttpField(String, String)` constructors.

Reply:
> As far as I can tell, every case can be replaced by a call to the `HttpField(HttpHeader, String)` or `HttpField(String, String)` constructors

They can be, but it will not be efficient.  If we have already looked up the cache to determine `HttpHeader` enum value from a name, then if we use `HttpField(String, String)` the CACHE will be looked up again.   But if we use 
`HttpField(HttpHeader, String)` we lose the ability to preserve any special case that an app might be (wrongly but still optionally supported) relying on.

I think the thing to do is to greatly reduce the calls to the 3 arg version and see what we have left.  I'll do a PR